src/06_ffi.md

```diff
@@ -228,17 +207,316 @@ library offers portable type aliases in `std:os::raw` (or `core::os::raw`):
 The [libc] crate offers more C compatible types that cover almost exhaustively
 the C standard library.
 
-> Rule {{#check FFI-PFTYPE | Use portable aliases `c_*` when binding to platform-dependent types}}
+> ### Rule {{#check FFI-PFTYPE | Use portable aliases `c_*` when binding to platform-dependent types}}
 >
 > In a secure Rust development, when interfacing with foreign code that
 > uses platform-dependent types, such as C's `int` and `long`, Rust code must
 > use portable type aliases, such as provided by the standard library or the
 > [libc] crate, rather than platform-specific types, except if
-> the binding is automatically generated for each platform (for instance,
-> with a [cbindgen] step in the build process).
+> the binding is automatically generated for each platform (see Note below).
+
+<!-- -->
+
+> **Note**
+>
+> Automatic binding generation tools (e.g. [cbindgen], [rust-bindgen]) are able
+> to ensure type consistency on a specific platform. They should be used during
+> the build process for each target to ensure that the generation is sound for
+> the specific target platform.
 
 [libc]: https://crates.io/crates/libc
 
+### Non-robust types: references, pointers, enums
+
+A *trap representation* of a particular type is a representation (pattern of
+bits) that respects the type's representation constraints (such as size and
+alignment) but does not represent a valid value of this type.
+
+A lot of Rust types have trap representations:
+
+- `bool` (1 byte, 256 representations, only 2 valid ones),
+- references,
+- pointers,
```

Comment:
I don't think that (raw) pointers have invalid bit-patterns; on the other hand _smart_ pointers do.

Reply:
# What we have

First, this "type robustness concept" is based on the definition of trap representation (notably used on the C specification): good representation with invalid value. Second, Rust memory model is not really defined yet, but for now it closely matches C when possible.

# Pointers robust?

## Pros

* For now a pointer can hold any value in practice.
* It makes clear that pointers are better for FFI than references.

## Cons

* Optimizations could use the fact that pointer are smaller 
* Having a pointer to a non-existing memory object is UB in C (except for a pointer just after an object). So at least there's at least one UB somewhere... and if Rust chooses to define an unsafe memory model one day maybe two!
* It's means an invalid pointer is a *valid value* of a pointer type >_<

I'm very conflicted about that, in my first draft I did not include pointers in the list...
I think I will wait on some return from the ANSSI guys.

Reply:
I think the ship has already sailed:

```rust
/// non-unsafe
fn any_ptr<T> (n: usize) -> *const T
{
    n as *const T
}
```

Reply:
Indeed it clearly shows that in Rust, **an invalid pointer is a valid value**